cleanlab/object_detection/rank.py

```diff
@@ -0,0 +1,1049 @@
+# This file is part of cleanlab.
+#
+# cleanlab is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# cleanlab is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with cleanlab.  If not, see <https://www.gnu.org/licenses/>.
+
+"""Methods to rank and score images in an object detection dataset (object detection data), based on how likely they
+are to contain label errors. """
+
+import warnings
+from typing import List, Dict, Any, Optional, Tuple, TypedDict
+import numpy as np
+import copy
+
+AuxiliaryTypesDict = TypedDict(
+    "AuxiliaryTypesDict",
+    {
+        "pred_labels": np.ndarray,
+        "pred_label_probs": np.ndarray,
+        "pred_bboxes": np.ndarray,
+        "lab_labels": np.ndarray,
+        "lab_bboxes": np.ndarray,
+        "similarity_matrix": np.ndarray,
+        "min_possible_similarity": float,
+    },
+)
+
+
+def get_label_quality_scores(
+    labels: List[Dict[str, Any]],
+    predictions: List[np.ndarray],
+    *,
+    method: str = "subtype",
+    probability_threshold: Optional[float] = None,
+    verbose: bool = True,
+) -> np.ndarray:
+    """Computes a label quality score for each image in the dataset.
+
+    For object detection datasets, the label quality score for an image estimates how likely it has been correctly labeled.
+    Lower scores indicate images whose annotation is more likely imperfect.
+    Annotators may have mislabeled an image because they:
+    - overlooked an object (missing annotated bounding box),
+    - chose the wrong class label for an annotated box in the correct location,
+    - imperfectly annotated the location/edges of a bounding box.
+    Any of these annotation errors should lead to an image with a lower label quality score.
+
+    Score is between 0 and 1.
+
+        - 1 - clean label (given label is likely correct).
+        - 0 - dirty label (given label is likely incorrect).
+
+    A score is calculated for each of ``N`` images, with ``K`` total classes in the data.
+    Each image has ``L`` annotated bounding boxes and ``M`` predicted bounding boxes.
+
+    Parameters
+    ----------
+    labels:
+        A list of ``N`` dictionaries such that ``labels[i]`` contains the given labels for the `i`-th image in the format
+       ``{'bboxes': np.ndarray((M,4)), 'labels': np.ndarray((M,)), 'image_name': str}`` where ``L`` is the number of annotated bounding boxes
+       for the `i`-th image and ``bboxes[j]`` is in the format ``[x,y,x,y]`` with given label ``labels[j]``. (``image_name`` is optional here)
```

Comment:
Do we want a helper function that will convert [x,y,x,y] to [x,y,w,h] format?
this will change a lot of functions, we can just force user to pass it in the current format

Reply:
We need to stick with one format throughout all of the cleanlab code. 

And then you can offer one helper function that user can use to change the format of their data, but we shouldn't be converting back-n-forth between different formats ever. So this conversion function should be the ONLY place in cleanlab codebase that the other data format is ever accepted as an input.

I also don't think adding a conversion function is high priority. First get this PR shipped, and then can add the conversion function in a follow-up PR.